### PR TITLE
gnrc.h: extend documentation

### PR DESCRIPTION
**Pull request by Lotterleben**
Add basic pktbuf usage to the tutorial in `gnrc.h` to draw attention to the fact that this should be done. Since the rest of the file doesn't mention the packet buffer, I think that it's important to point out its existence, since it could easily be overlooked otherwise.

Additionally, in some instances, I've found the text I read a bit hard to understand, so I've clarified the language.

Disclaimer: I've just written down what I think is correct from the code I've read, so I'm not 100% confident I got it right ;)

### CODE REVIEW

File: sys/include/net/gnrc.h [hunk 1]
```diff
@@ -155,26 +155,46 @@
  * Transmitting Packets
  * --------------------
  *
- * To relay the @ref net_gnrc_pkt to interested threads the @ref
- * net_gnrc_netapi offers a dispatch functionality. The following example
+ * A packet is transmitted by relaying it to threads interested in handling (and
+ * dispatching) packets of its type. To do this, the @ref
+ * net_gnrc_netapi offers a dispatch functionality called @ref
+ * gnrc_netapi_dispatch_send().
```
Comment (miri64) [inline sys/include/net/gnrc.h]: `.. and gnrc_netapi_dispatch_receive()`. Also I would prefer that it is pointed out that these are only helper functions. One can as well use raw `msg_send()` if they know what they are doing.

File: sys/include/net/gnrc.h [hunk 2]
```diff
@@ -155,26 +155,46 @@
...
+ *
+ * The following example
```
Comment (miri64) [inline sys/include/net/gnrc.h]: `Jgqq` (vim2human: please join and with following line and truncate to 72 chars again please)

File: sys/include/net/gnrc.h [hunk 3]
```diff
@@ -155,26 +155,46 @@
...
  *      puts("Error: no thread is interested");
  *      gnrc_pktbuf_release(pkt);
  *      return;
  * }
  * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  *
- * The @ref net_gnrc_pkt "pkt" will be send to all threads that registered for
- * @ref GNRC_NETTYPE_UDP and the demux context `80`. Every registered thread
+ * First, the data to be sent is added to the packet buffer. This ensures its
+ * intactness during the sending process. After the data in @data has been added
```
Comment (miri64) [inline sys/include/net/gnrc.h]: Is "intactness" a word?

File: sys/include/net/gnrc.h [hunk 4]
```diff
@@ -155,26 +155,46 @@
...
  * if (!gnrc_netapi_dispatch_send(GNRC_NETTYPE_UDP, 80, pkt)) {
  *      puts("Error: no thread is interested");
  *      gnrc_pktbuf_release(pkt);
  *      return;
  * }
  * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  *
- * The @ref net_gnrc_pkt "pkt" will be send to all threads that registered for
- * @ref GNRC_NETTYPE_UDP and the demux context `80`. Every registered thread
+ * First, the data to be sent is added to the packet buffer. This ensures its
```
Comment (miri64) [inline sys/include/net/gnrc.h]: `is added to the @ref net_gnrc_pktbuf "packet buffer"`.

File: sys/include/net/gnrc.h (hunk 3 above)
Comment (miri64) [inline sys/include/net/gnrc.h]: Don't know what `@data` is supposed to mean? Maybe `gnrc_pktsnip_t::data`?

File: sys/include/net/gnrc.h [hunk 5]
```diff
@@ -155,26 +155,46 @@
...
+ * for @ref GNRC_NETTYPE_UDP and the demux context `80`. Every registered thread
  * will receive a @ref GNRC_NETAPI_MSG_TYPE_SND command and can access the @ref
  * net_gnrc_pkt. If @ref gnrc_netapi_dispatch_send() is replaced by @ref
  * gnrc_netapi_dispatch_receive() then threads will receive the @ref
  * GNRC_NETAPI_MSG_TYPE_RCV command instead, again with access to the @ref
  * net_gnrc_pkt.
  *
+ * @note The IP and UDP headers for pkt have to be built manually before
+ *       calling @ref gnrc_netapi_dispatch_send(). In the interest of conciseness,
+ *       this is omitted in this tutorial- please refer to @ref gnrc_udp_hdr_build(),
```
Comment (miri64) [inline sys/include/net/gnrc.h]: `s/-/;/`?

File: sys/include/net/gnrc.h [hunk 6]
```diff
@@ -155,26 +155,46 @@
...
+ *
+ * Then, the @ref net_gnrc_pkt "pkt" will be sent to all threads that registered
+ * for @ref GNRC_NETTYPE_UDP and the demux context `80`. Every registered thread
  * will receive a @ref GNRC_NETAPI_MSG_TYPE_SND command and can access the @ref
  * net_gnrc_pkt. If @ref gnrc_netapi_dispatch_send() is replaced by @ref
  * gnrc_netapi_dispatch_receive() then threads will receive the @ref
  * GNRC_NETAPI_MSG_TYPE_RCV command instead, again with access to the @ref
  * net_gnrc_pkt.
  *
+ * @note The IP and UDP headers for pkt have to be built manually before
```
Comment (miri64) [inline sys/include/net/gnrc.h]: `have to be` gives the wrong idea IMHO. This is only the fact if you want to send to the UDP module.

File: sys/include/net/gnrc.h [hunk 7]
```diff
@@ -155,26 +155,46 @@
...
  *
  * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * gnrc_pktsnip_t *pkt;
+ *
+ * pkt = gnrc_pktbuf_add(NULL, data, size, GNRC_NETTYPE_UNDEF);
+ * if (pkt == NULL) {
+ *      puts("Error: unable to copy data to packet buffer\n");
+ *      return;
+ * }
  * if (!gnrc_netapi_dispatch_send(GNRC_NETTYPE_UDP, 80, pkt)) {
```
Comment (miri64) [inline sys/include/net/gnrc.h]: Maybe add a comment, that the own-age of `pkt` is given to the receiver threads by calling this function?

File: sys/include/net/gnrc.h [hunk 8]
```diff
@@ -144,9 +144,9 @@
  * }
  * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  *
- * @note It is necessary to react with a reply message to the other thread for
- *       @ref GNRC_NETAPI_MSG_TYPE_SET and @ref GNRC_NETAPI_MSG_TYPE_GET
- *       commands
+ * @note When receiving a message of type @ref GNRC_NETAPI_MSG_TYPE_SET or
+ *       @ref GNRC_NETAPI_MSG_TYPE_GET, it is necessary to acknowledge it by
+ *       calling @ref msg_reply()
```
Comment (Lotterleben) [inline sys/include/net/gnrc.h]: good point, done :)

File: sys/include/net/gnrc.h (hunk 2 above)
Comment (Lotterleben) [inline sys/include/net/gnrc.h]: :D done.

File: sys/include/net/gnrc.h (hunk 7 above)
Comment (Lotterleben) [inline sys/include/net/gnrc.h]: Good idea. Quick question though: my editor (sublimetext) gets *really* confused when I add RIOT-style comments (i.e. `/* ... */`) there, which makes the rest of the file a pain to read when I use them. Do we have a policy for such nested comments? Am I allowed to use `//` or should I just explain the whole thing later in the text?

File: sys/include/net/gnrc.h (hunk 4 above)
Comment (Lotterleben) [inline sys/include/net/gnrc.h]: good catch, thanks, done :)

File: sys/include/net/gnrc.h (hunk 6 above)
Comment (Lotterleben) [inline sys/include/net/gnrc.h]: True. How about... "If the data to be sent requires extra headers to be added for successful transmission (in the example, this would be IP and UDP headers), these have to be built manually ..."?

File: sys/include/net/gnrc.h (hunk 5 above)
Comment (Lotterleben) [inline sys/include/net/gnrc.h]: allrighty :)

File: sys/include/net/gnrc.h (hunk 7 above)
Comment (miri64) [inline sys/include/net/gnrc.h]: Ah there is not only a policy for that, that's illegal in C... I think explaining this later is the better choice here... :-/

File: sys/include/net/gnrc.h (hunk 1 above)
Comment (miri64) [inline sys/include/net/gnrc.h]: I don't think it's confusing to replace `a dispatch functionality` with `a dispatch helper function` ;-).

File: sys/include/net/gnrc.h (hunk 7 above)
Comment (Lotterleben) [inline sys/include/net/gnrc.h]: makes sense :D I'll fix that as soon as I get home, thanks for all of your reviews :)

File: sys/include/net/gnrc.h (hunk 3 above)
Comment (miri64) [inline sys/include/net/gnrc.h]: Because there is no (public) variable called data (at least in GNRC). Which data do you want to reference?

File: sys/include/net/gnrc.h (hunk 6 above)
Comment (miri64) [inline sys/include/net/gnrc.h]: :+1:

File: sys/include/net/gnrc.h [hunk 9]
```diff
@@ -144,9 +144,10 @@
  * }
  * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  *
- * @note It is necessary to react with a reply message to the other thread for
- *       @ref GNRC_NETAPI_MSG_TYPE_SET and @ref GNRC_NETAPI_MSG_TYPE_GET
- *       commands
+ * @note When receiving a message of type @ref GNRC_NETAPI_MSG_TYPE_SET or
+ *       @ref GNRC_NETAPI_MSG_TYPE_GET, it is necessary to acknowledge it by
+ *       calling @ref msg_reply() with a message of type
+ *       @ref GNRC_NETAPI_MSG_TYPE_ACK.
```
Comment (miri64) [inline sys/include/net/gnrc.h]: Now reading this with my newbie hat, I'm not sure if at least mentioning that the value field of that message must also be set to the return value is helpful or not...

File: sys/include/net/gnrc.h (hunk 6 above)
Comment (Lotterleben) [inline sys/include/net/gnrc.h]: done :)

File: sys/include/net/gnrc.h (hunk 1 above)
Comment (Lotterleben) [inline sys/include/net/gnrc.h]: Oooh now I get it. Yeah, done :D

File: sys/include/net/gnrc.h (hunk 3 above)
Comment (Lotterleben) [inline sys/include/net/gnrc.h]: the data being sent, i.e. the `gnrc_pktbuf_add()` input called data :D I thought it'd make sense to differentiate between "some arbitrary data" and "data, the input", but actually it doesn't really, does it?

TL;DR: nevermind, I'll delete the `@`. :D

File: sys/include/net/gnrc.h (hunk 9 above)
Comment (Lotterleben) [inline sys/include/net/gnrc.h]: As a newbie I think that's very helpful, but I'm not sure I've correctly understood how it works yet.. By looking at how `gnrc_zep.c:_get()` and `gnrc_zep.c:_event_loop()` work and `gnrc_netapi_get()` is documented, I'd think that the content of the reply message should be either an error code or the size of the content of the message that triggered the reply. Is that the case?

Comment (miri64) [inline sys/include/net/gnrc.h]: > By looking at how gnrc_zep.c:_get() and gnrc_zep.c:_event_loop() 

Please don't look at ZEP... I did not touch it for ages and it wasn't working then either ;-) Better look at `gnrc_netdev2` ;-).

> gnrc_netapi_get() is documented, I'd think that the content of the reply message should be either an error code or the size of the content of the message that triggered the reply. Is that the case?

Error code is correct. For success it returns the size of the returned value (which need not be the same value as `max_len` given by the triggering message).

Comment (Lotterleben) [inline sys/include/net/gnrc.h]: allright, but what exactly is "the returned value"? the value returned by the function that was called in reaction to the receipt of the reply-triggering message?

Also, I'm wondering if it makes sense to additionally document this somewhere more prominent than that tutorial ^^

Comment (miri64) [inline sys/include/net/gnrc.h]: The value that is at `gnrc_netapi_opt_t::data` (which in turn is pointed to by `msg.content.ptr` of the GNRC_NETAPI_MSG_TYPE_GET msg) on reply. Since that is a pointer this information needs to be provided to the sender somehow (gnrc_netapi_opt_t::data_len only provides the replier with the *maximum space* available at `gnrc_netapi_opt_t::data`).

Comment (Lotterleben) [inline sys/include/net/gnrc.h]: Oh, that makes a lot of sense, thanks! I've changed it to
"[...] with a message of type @ref GNRC_NETAPI_MSG_TYPE_ACK which contains the actual size of the GET message's content on success or an error code otherwise."

now... But I'm wondering, this is quite a tricky thing to explain & get right, where would be the correct place to mention this in detail?

**Comment by miri64** [comment]
ACK, please squash.

**Comment by Lotterleben** [comment]
Squashed.
